### PR TITLE
Musl support

### PR DESCRIPTION
**Pull request by gnurizen**
Add support for pulling unwind info from debug link that is available
locally, if debug link can't be found dwarf unwinding will fail. Future work
may have the agent to pull debuginfo from parca server and allow
dwarf unwinding w/o debug link files available locally.

If the debug link file doesn't match the CRC report an error and don't
use it.

Fix finder so that it doesn't return debuglink matches pointing back at
the exe itself, this just leads to spurious crc failures.

### CODE REVIEW

File: pkg/objectfile/object_file.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ import (
 // ObjectFile represents an executable or library file.
 // It handles the lifetime of the underlying file descriptor.
 type ObjectFile struct {
-	p *Pool
+	Pool *Pool
```
Comment (kakkoyun) [inline pkg/objectfile/object_file.go:33]: Why do we export this?

File: pkg/stack/unwind/unwind_table.go [hunk 2]
```diff
@@ -200,37 +203,84 @@ func (ptb *UnwindTableBuilder) PrintTable(writer io.Writer, file *objectfile.Obj
 }
 
 func ReadFDEs(file *objectfile.ObjectFile) (frame.FrameDescriptionEntries, elf.Machine, error) {
+	// Try the normal path, if it fails try the alternatives and go with them if they
```
Comment (kakkoyun) [inline pkg/stack/unwind/unwind_table.go:206]: It worth checking our debuginfo package. Especially the `Find`. I believe we can reuse some code pieces.

File: pkg/objectfile/object_file.go (hunk 1 above)
Comment (gnurizen) [reply pkg/objectfile/object_file.go]: Ignore, WIP temporary hack.

File: pkg/stack/unwind/unwind_table.go (hunk 2 above)
Comment (gnurizen) [reply pkg/stack/unwind/unwind_table.go:227]: Thanks for the tip!  I especially want to nail the CRC checking before landing this.

**Comment by gnurizen** [comment]
@umanwizard  can you review this?   Thanks!

**Review by brancz — APPROVED** [review]

File: bpf/unwinders/native.bpf.c [hunk 3]
```diff
@@ -688,8 +691,9 @@ static __always_inline enum find_unwind_table_return find_unwind_table(chunk_inf
         }
     }
 
-    LOG("[error] could not find chunk for adjusted ip=0x%llx", adjusted_pc);
-    return FIND_UNWIND_CHUNK_NOT_FOUND;
+    LOG("[error] could not find chunk for adjusted ip=0x%llx, mapping idx %d, mapping exe id 0x%llx", adjusted_pc, index,
+        proc_info->mappings[index].executable_id);
```
[the remark this answers is not in the document]
Comment (gnurizen) [reply bpf/unwinders/native.bpf.c:695]: Will do

File: bpf/unwinders/native.bpf.c [hunk 4]
```diff
@@ -1402,7 +1411,7 @@ int entrypoint(struct bpf_perf_event_data *ctx) {
         if (!is_debug_enabled_for_thread(per_process_id)) {
             bump_unwind_total_filter_misses();
             BUMP_UNWIND_FAILED_COUNT(per_process_id, missed_filter);
-            LOG("[debug] pid %u didn't match filter, ignoring.", per_process_id);
+            // LOG("[debug] pid %u didn't match filter, ignoring.", per_process_id);
```
[the remark this answers is not in the document]
Comment (gnurizen) [reply bpf/unwinders/native.bpf.c:1414]: Too verbose

File: pkg/stack/unwind/compact_unwind_table.go [hunk 5]
```diff
@@ -158,6 +152,14 @@ func BuildCompactUnwindTable(fdes frame.FrameDescriptionEntries, arch elf.Machin
 	context := frame.NewContext()
 	lastFunctionPc := uint64(0)
 	for _, fde := range fdes {
+		// We use pc=0 as a sentinel so we can't have that, however some debug
+		// files have FDE's with offset 0, usually empty or tiny but not always.
+		// There's probably a better way to filter these but this works, we
+		// panic over in maps.setUnwindTableForMapping if any start at 0.
+		// An example is the alpine "ld-musl-x86_64.so.1.debug".
```
[the remark this answers is not in the document]
Comment (gnurizen) [reply pkg/stack/unwind/compact_unwind_table.go]: I tried to repro it and couldn't, I wonder if it was from parsing .debug_frame wrong.   We'll see what more testing shows.

File: pkg/stack/unwind/unwind_table.go [hunk 6]
```diff
@@ -199,37 +212,178 @@ func (ptb *UnwindTableBuilder) PrintTable(writer io.Writer, file *objectfile.Obj
...
+			fdes = nil
+		}
+		debugFrameFDEs = fdes
+	}
+
+	// Add FDEs from debuglink, this is necessary for stripped binaries with no frame pointers
+	// and external debuginfo's.
+	var debugLinkFDEs frame.FrameDescriptionEntries
+	if uc.finder != nil {
+		debugPath, err := uc.finder.Find(context.TODO(), root, file)
```
[the remark this answers is not in the document]
Comment (gnurizen) [reply pkg/stack/unwind/unwind_table.go:273]: Well we should have a legit context here but we don't, this is a bread crumb to remember to wire in a real context.

**Comment by umanwizard** [comment]
I still don’t have approval rights, but it looks good to me overall, so merge at your discretion!

File: pkg/stack/unwind/unwind_table.go [hunk 7]
```diff
@@ -199,37 +212,178 @@ func (ptb *UnwindTableBuilder) PrintTable(writer io.Writer, file *objectfile.Obj
...
+		if err != nil {
+			level.Debug(uc.logger).Log("msg", "no .debug file found for ", "exe", exe, "err", err)
+		} else if debugPath != "" {
```
[the remark this answers is not in the document]
Comment (gnurizen) [reply pkg/stack/unwind/unwind_table.go:276]: Sure

File: pkg/stack/unwind/unwind_table.go [hunk 8]
```diff
@@ -199,37 +212,178 @@ func (ptb *UnwindTableBuilder) PrintTable(writer io.Writer, file *objectfile.Obj
...
+			fdes, arch3, err := readFDEsFromSection(obj, ".debug_frame")
 
-	var secName string
-	if ehFrame {
-		secName = ".eh_frame"
-	} else {
-		secName = ".debug_frame"
+			if err != nil {
+				if isUnexpected(err) {
+					level.Warn(uc.logger).Log("msg", "error reading FDEs from .ebug_frame section in debuglink", "err", err, "exe", exe, "debuglink", debugPath)
```
[the remark this answers is not in the document]
Comment (gnurizen) [reply pkg/stack/unwind/unwind_table.go:291]: Nice catch!

File: bpf/unwinders/native.bpf.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (gnurizen) [reply bpf/unwinders/native.bpf.c:695]: It seems to work on ubuntu 20.04 LTS (5.14 kernel).  Are you sure you weren't just fixing printfs with more than 3 args?

**Pull request merged**